### PR TITLE
Add management command to track Kudos purchases

### PR DESCRIPTION
**Pull request by santteegt**
##### Description

Gitcoin team members want to track kudos purchases by recording revenue in the database. So, a management command to track Kudos purchases is implemented to track ETH and token transaction from a specified account address, while a cron job is setup to run the command once an hour.

##### Checklist

- [X] Read and conforms to the [contributor guidelines](https://docs.gitcoin.co/mk_contributors/)
- [X] Read and conforms to the [style guidelines](https://docs.gitcoin.co/mk_styleguide/)
- [X] Linter status: 100% pass
- [x] Added relevant tests
- [x] Tests pass and code coverage has not decreased
- [X] Changes don't break existing behavior
- [X] Commit message follows [commit guidelines](https://docs.gitcoin.co/mk_contributors/#step-4-commit)
- [ ] [Code documentation](https://docs.gitcoin.co/mk_contributors/#docstrings)
- [ ] Completed: Code review by core team and any requested changes

##### Affected core subsystem(s)

* revenue

##### Refers/Fixes

Refs: #4122 

##### Testing and Sign-off

Before testing new features please create an Etherscan API Key on https://etherscan.io/apis and set its value on `ETHERSCAN_API_KEY` env variable.

###### Contributor

- [X] Read and followed the [Contributor Guidelines](https://docs.gitcoin.co/mk_contributors/)
- [X] Tested all changes **locally**
- [X] Verified existing functionality
- [x] Ran `make test` and everything passed!

###### Reviewer

- [ ] Affirm contributor guidelines have been followed and requested changes made
- [ ] CI tests and linting pass
- [ ] No conflicts (migrations, files, etc)
- [ ] Regression tested against staging or local deployment

###### Funder

- [ ] Validated requested changes were made to specification
- [ ] Bounty payout released to the contributor

### CODE REVIEW

File: scripts/crontab [hunk 1]
```diff
@@ -19,6 +19,7 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/us
 ## KUDOS
 */15 3,10,17 * * * cd gitcoin/coin; bash scripts/run_management_command_if_not_already_running.bash sync_kudos mainnet opensea --catchup >> /var/log/gitcoin/sync_kudos_catchup.log 2>&1
 */10 4,11,18 * * * cd gitcoin/coin; bash scripts/run_management_command_if_not_already_running.bash sync_kudos mainnet opensea --start 1 >> /var/log/gitcoin/sync_kudos_all.log 2>&1
+0 */1 * * * cd gitcoin/coin; bash scripts/run_management_command_if_not_already_running.bash kudos_revenue mainnet --account-address 0xdb282cee382244e05dd226c8809d2405b76fbdc9
```
Comment (owocki) [inline scripts/crontab]: *1 == * right?

File: app/revenue/management/commands/kudos_revenue.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        print(f"Total {params['action']} found in blocks {params['startblock']} to {params['endblock']}: {len(records)}")
+        for rec in records:
+            save_incoming_tx(network, rec)
+    except Exception as e:
+        logger.error(f"* Exception in kudos_revenue => {e}")
+
+
+def save_incoming_tx(network, record):
+    decimals = 18. if len(record['contractAddress']) == 0 else float(record['tokenDecimal'])
+    dgp = DigitalGoodPurchase.objects.create(
```
Comment (owocki) [inline app/revenue/management/commands/kudos_revenue.py]: what if this specific DGP already exists?!?

File: app/revenue/management/commands/kudos_revenue.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        rs = requests.get(api_url, params).json()
+        records = rs['result']
+        if rs['message'] != 'OK':
+            if len(records) == 0:
+                logger.info(f"No {params['action']} found in blocks {params['startblock']} to {params['endblock']}")
+            else:
+                logger.error(f"* Error during API call => {rs['message']}")
+            return
+        print(f"Total {params['action']} found in blocks {params['startblock']} to {params['endblock']}: {len(records)}")
+        for rec in records:
```
Comment (owocki) [inline app/revenue/management/commands/kudos_revenue.py]: seems like the records should be turned from this method and save_incoming_tx() should be called outside of this scope..

just to keep each function small + cohesive

File: app/revenue/management/commands/kudos_revenue.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            .filter(receive_address__iexact=options['account_address'], tokenName__iexact='ETH') \
+            .order_by('-id') \
+            .values('id', 'metadata')[:1]
+        
+        if len(dgp_records) > 0:
+            params['startblock'] = int(dgp_records[0]['metadata']['blockNumber']) + 1
+            logger.info(f"ETH Tx Records were found. Continuing from block {params['startblock']}")
+        
+        params['action'] = 'txlist'
+        call_etherscan_api(network, params)
```
Comment (owocki) [inline app/revenue/management/commands/kudos_revenue.py]: why do we need to call `call_etherscan_api()` twice?  why not just once

Comment (santteegt) [reply app/revenue/management/commands/kudos_revenue.py]: Etherscan API uses two separate methods to retrieve ETH transactions (txlist) and ERC-20 token transactions (tokentx)

File: scripts/crontab (hunk 1 above)
Comment (santteegt) [reply scripts/crontab]: https://crontab.guru/#0_*/1_*_*_*

File: app/revenue/management/commands/kudos_revenue.py (hunk 3 above)
Comment (santteegt) [reply app/revenue/management/commands/kudos_revenue.py]: my bad! I Just Fixed it!

File: app/revenue/management/commands/kudos_revenue.py (hunk 2 above)
Comment (santteegt) [reply app/revenue/management/commands/kudos_revenue.py]: Fixed!

**Comment by owocki** [comment]
this LGTM.  @danlipert look ok to you?

**Comment by PixelantDesign** [comment]
bumping this @danlipert ^^

File: app/revenue/management/commands/kudos_revenue.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+from django.core.management.base import BaseCommand
+
+import requests
+from revenue.models import DigitalGoodPurchase
+
+warnings.filterwarnings("ignore", category=DeprecationWarning)
+logging.getLogger("requests").setLevel(logging.WARNING)
+logger = logging.getLogger(__name__)
+
+REVENUE_ACCOUNT_ADDRESS = '0xdb282cee382244e05dd226c8809d2405b76fbdc9'
```
Comment (danlipert) [inline app/revenue/management/commands/kudos_revenue.py]: Lets move this to settings

**Comment by danlipert** [comment]
@santteegt Can you add some tests for this? Thanks!

**Comment by santteegt** [comment]
Hi @danlipert,

I've just updated my PR to consider your review comments

File: app/app/settings.py [hunk 6]
```diff
@@ -516,6 +516,12 @@
 GITHUB_API_TOKEN = env('GITHUB_API_TOKEN', default='')  # TODO
 GITHUB_APP_NAME = env('GITHUB_APP_NAME', default='gitcoin-local')
 
+# Etherscan API
+ETHERSCAN_API_KEY = env('ETHERSCAN_API_KEY', default='')
+
+# Kudos revenue account
+KUDOS_REVENUE_ACCOUNT_ADDRESS = env('KUDOS_REVENUE_ACCOUNT_ADDRESS', default='0xdb282cee382244e05dd226c8809d2405b76fbdc9')
```
Comment (danlipert) [inline app/app/settings.py]: @owocki is this the correct address (and is it the same on rinkeby, or is there even a rinkeby equivalent?)

**Comment by danlipert** [comment]
@santteegt looking great - I'm wondering if you can test the management command directly and actually inspect the resulting DGP objects that are created after running to ensure they match whats on the blockchain. Also, how are the tests returning data properly even though no API key is set? 🤔

**Comment by santteegt** [comment]
Hi @danlipert,

I've finally added a test case that directly performs a call to the management command. Regarding the empty API key during tests, Etherscan has a limited quota to call their API for free without an API_KEY

**Comment by danlipert** [comment]
@owocki how is this looking to you?

File: app/app/settings.py (hunk 6 above)
Comment (owocki) [reply app/app/settings.py]: no rinkeby equiv; this is the right addr yes

**Comment by danlipert** [comment]
API Key is in place for when this gets merged 👍

**Comment by danlipert** [comment]
@santteegt we are almost able to get this in, but the automated tests are failing because of the reliance on Etherscan. Can you mock those out so that our automated tests don't depend on Etherscan's availability or whether they block the free requests in the future? Thanks!

**Comment by santteegt** [comment]
@danlipert sure! I'll try to get this done ASAP

**Comment by thelostone-mc** [comment]
@santteegt any update on this . ?

**Comment by santteegt** [comment]
Hi @thelostone-mc,

Sorry for the late PR submission. I just updated it for you to review

**Pull request merged**